File: 2015/day2.py
import AOCInit
import util

inp = util.getInput(d=2, y=2015)
arr = list(list(int(c) for c in l.split('x')) for l in inp.splitlines())

def paperSize(side):
    p = util.prod(side)
    m = max(side)
    return p / m + sum(p / s for s in side) * 2

def ribbonSize(side):
    s = sum(side)
    m = max(side)
    return 2 * (s - m) + util.prod(side)

# part 1
print(sum(paperSize(l) for l in arr))

# part 2
print(sum(ribbonSize(l) for l in arr))

File: 2015/day2.rb
inp = File.read('input2').split.map { |line| line.scan(/\d+/).map(&:to_i) }

# part 1
puts inp.map { |arr| arr.sum**2 - arr.map { |x| x**2 }.sum + arr.inject(:*) / arr.max }.sum

# part 2
puts inp.map { |arr| 2 * (arr.sum - arr.max) + arr.inject(:*) }.sum
